feat(#61): order status update api

joel-you opened the pull request:
### 🔗 관련 이슈

- Related to #61

### 📝 변경 사항

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 리팩토링
- [ ] 성능 개선
- [ ] 문서 업데이트
- [x] 테스트 코드 추가/수정
- [ ] 기타:

### 🛠️ 기술적 변경 사항

<!-- 기술적인 변경 내용을 자세히 설명해 주세요 -->

Domain Layer (3개 커밋)
- Order 엔티티에 changeStatus() 메서드 추가 (상태 전이 검증)
- 상태별 타임스탬프 자동 업데이트
- 도메인 테스트 23개 추가

Application Layer (2개 커밋)
- ChangeOrderStatusUseCase 인터페이스 및 서비스 구현
- Command/Result DTO 생성
- 단위 테스트 6개 추가 (Mockito)

Adapter Layer (4개 커밋)
- PATCH /api/v1/orders/{orderId}/status 엔드포인트 추가
- Request/Response DTO 및 MapStruct 매퍼
- GlobalExceptionHandler에 IllegalStateException 처리 추가
- 통합 테스트 7개 추가

### 🧪 테스트

- [x] 단위 테스트 통과
- [x] 통합 테스트 통과
- [ ] E2E 테스트 통과 (해당하는 경우)
- [ ] 수동 테스트 완료

### 🔒 보안 체크리스트

- [ ] 새로운 의존성 추가 시 보안 검토 완료
- [x] 사용자 입력 검증 로직 확인
- [ ] 인증/권한 로직 검토 완료
- [x] 민감한 정보 노출 위험 없음

### 🔄 데이터베이스 변경

- [x] 데이터베이스 스키마 변경 없음
- [ ] Migration 파일 추가됨
- [ ] 기존 데이터와 호환됨
- [ ] 롤백 계획 수립됨

### 📱 호환성

- [x] 이전 버전과 호환됨
- [x] Breaking Change 없음
- [ ] API 변경 없음
- [ ] 환경 설정 변경 필요 (상세: )

### 📖 문서 업데이트

- [ ] README.md 업데이트 완료
- [ ] API 문서 업데이트 완료
- [ ] 코드 주석 업데이트 완료
- [ ] 별도 문서 업데이트 불필요

### ✅ 체크리스트

- [x] 코드가 컨벤션을 따르고 있습니다
- [x] 자가 리뷰를 완료했습니다
- [ ] 적절한 라벨을 추가했습니다
- [ ] 관련 문서를 업데이트했습니다
- [x] 테스트를 추가/업데이트했습니다
- [x] CI/CD 파이프라인이 통과합니다

joel-you committed:
feat(#61): 주문 상태 전이 로직 구현

- Order 엔티티에 changeStatus 메서드 추가
- 상태 전이 검증 (PENDING→CONFIRMED/CANCELLED 등)
- 타임스탬프 자동 업데이트

joel-you committed:
feat(#61): 주문 상태 변경 서비스 구현

ChangeOrderStatusUseCase 및 서비스 구현
- Command/Result DTO 생성
- 트랜잭션 관리 및 주문 존재 검증

joel-you committed:
feat(#61): 주문 상태 변경 API 구현

PATCH /api/v1/orders/{orderId}/status 엔드포인트 추가
- OrderCommandController 및 DTO 구현
- MapStruct 매퍼로 레이어 간 변환
- GlobalExceptionHandler에 IllegalStateException 처리 추가

joel-you committed:
docs(#61): 주문 상태 변경 API 문서 추가

- 상태 전이 규칙 및 비즈니스 로직 문서화
- OrderStatusChangeRequest/Response 스키마 정의
- 요청/응답 예제 3개씩 추가
- 에러 케이스 문서화 (400, 404, 500)

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/web/OrderCommandController.java:
```diff
@@ -0,0 +1,45 @@
+        @PathVariable UUID orderId,
+        @Valid @RequestBody OrderStatusChangeRequest request
+    ) {
+        ChangeOrderStatusCommand command = orderStatusChangeMapper.toCommand(orderId, request);
```
외부에서 상태 전이에 대한 이후 값을 받고, 도메인(Order)에서 검증을 하는 것으로 이해했습니다. 
다음 단계로만 진행되는 프로세스이고, 분기가 없는 것으로 보이는데요. 상태 값을 외부에서 받아서 변경하는 방식보다, 도메인 안에서 행위를 호출하여 다음 상태로 계산하는 방식으로 고려해 주시겠어요?

joel-you replied:
수정하겠습니다!

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/web/dto/request/OrderStatusChangeRequest.java:
```diff
@@ -0,0 +1,18 @@
+public record OrderStatusChangeRequest(
+
+    @NotNull(message = "New status is required")
+    OrderStatus newStatus,
```
도메인 엔티티(또는 Aggregate) 등 에서 전기 규칙 캡슐화하여 도메인 내부 행위로 전이되면 좋겠어요.

joel-you replied:
해당 DTO가 제거되었고, 다른 로직에서 적용해보겠습니다

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/web/OrderCommandController.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final OrderStatusChangeMapper orderStatusChangeMapper;
+
+    @PatchMapping("/{orderId}/status")
```
각 상태별로 여러 비즈니스 행위들이 존재할 것 같습니다. 행위 기반 엔드 포인트를 분리하면 좋겠어요. 
```java
@PostMapping("/{orderId}/confirm")
@PostMapping("/{orderId}/cancel")
@PostMapping("/{orderId}/paid")
```

joel-you replied:
미팅 때 이 얘기에 대해서 의논이 되었던 것으로 기억하는데요.

저희 회사에서는 상태값 전이에 대해서 하나의 API 로 관리(기존 방식)된다고 말씀 드렸었는데, 핵심 비즈니스라고 생각해보니 행위 기반으로 나누어져 있는게 있었네요.

따라서 Order라는 핵심 비즈니스에서 행위기반으로 나누는게 더 올바르다고 생각해서 수정해 보겠습니다.

flab-true commented on docs/api-specs/order-orchestrator-openapi.yml:
```diff
@@ -35,8 +41,145 @@ tags:
+      description: |
+        주문 상태를 변경합니다.
+
+        ### Business Rules
```
Strategy(=Policy) 패턴 vs State 패턴 vs Template Method 패턴에 대한 이야기를 나눠보고 싶어요.

joel-you replied:
논의된 내용을 바탕으로 State 패턴을 도입했습니다.
하고나니까 변형된 State Pattern 이 된 것 같은데, over engineering 이 된 것 같기도하네요 ..

다른 패턴을 도입하지 않은 이유는 다음과 같습니다.
* 전략패턴
    * 상태 관리에는 부적합 한 것 같음
    * 전략의 경우 "방법"을 캡슐화 하는 것이 목적인 것 같음
    * 상태 간 관계 표현 어려워 보임
    * 클라이언트 책임 증가
        * Order가 적절한 전략을 선택해야 함
* Template Method Pattern
    * PENDING→CONFIRMED, CONFIRMED→PAYMENT_PENDING 등 각각 클래스가 State 전략보다 많이 필요한 듯함
    * 상속의 한계가 있어서 유연성이 감소함
    * 현재 상태에서 무엇을 할 수 있는가 표현이 어려움

flab-true replied:
@joel-you 님 고려해 주셔서 감사합니다. 
행위 기반으로 API를 변경해 주셨고, 관련 신규 Comman 인해 Class가 많이 생성되었네요. 
Order 에서 상태에 대한 설계는 매우 중요합니다. State 패턴을 조금 더 들어가 보겠습니다.
상태 전이를 공부하시면 `SMM`, `FSM` 용어들이 나옵니다. 

FSM을 예로 들면, 
주로 컴퓨터 프로그램이나 전자 회로 설계할 때 개수가 한정된 상태와 이동을 트리거하는 조건들의 집합으로 정의하는데요.
즉 FSM 디자인 패턴은 정해져있는 유한한 상태와 전이를 정의하는 트리거를 정의해 상태를 관리하는 디자인 패턴입니다. 
쉽게 말해 State Pattern의 next prev가 아닌 input으로 받는 값에 따라 상태를 바꾸는 것이라고 보시면 됩니다.
Spring 에서는 이를 지원하는 라이브러리가 존재합니다. spring-statemachine-core 를 한번 검색해 보시면 좋겠습니다.

flab-true reviewed (commented):
상태 전이에 대한 코드를 작성해 주셔서 감사합니다. 상태 전이에 대해서 이야기를 나눠보고 싶어요.
간단한 코멘트를 남깁니다.

joel-you committed:
refactor(#61): OrderTest를 행위 기반 메서드로 리팩토링

- changeStatus() 호출을 confirm(), cancel() 등으로 변경
- 더 이상 존재하지 않는 changeStatus() 테스트 제거
- 도메인 규칙 위반 테스트 제거 (PROCESSING->CANCELLED)

joel-you committed:
refactor(#61): Order 상태 관련 로직 리팩토링

- State Pattern 으로 리팩토링 진행함
- api endpoint 와 맞게 method name 변경
- Setter를 의미있는 메소드명으로 변경

joel-you commented:
수정 완료

flab-true reviewed (approved):
LGTM 👍 
현업에서 Dev Talk 시간을 구성해 보시고, State 에 대한 고민을 공유하는 자리를 마련하셔도 좋겠습니다. 좋은 에너지를 동료들에게 전달해 주시죠.
